src/runtime/hexagon_cache_allocator.cpp

```diff
@@ -0,0 +1,165 @@
+#include "HalideRuntime.h"
+#include "mini_hexagon_dma.h"
+#include "HalideRuntimeHexagonDma.h"
+
+namespace Halide { namespace Runtime { namespace Internal {
+
+typedef struct hexagon_cache_block {
+    void *l2memory;
+    size_t bytes;
+    bool used;
+    struct hexagon_cache_block *next;
+} hexagon_cache_pool_t;
+
+typedef hexagon_cache_pool_t *pcache_pool;
+
+pcache_pool hexagon_cache_pool = NULL;
+halide_mutex hexagon_cache_mutex; 
+
+}}}
+
+using namespace Halide::Runtime::Internal;
+
+extern "C" {
+//TO DO Add hjalide_mutex_lock to make it thread sahe 
+static inline void *hexagon_cache_pool_get (void *user_context, size_t size, bool retry) {
+    // TODO: Add Mutex locking for access to hexagon_cache_pool ( To be thread safe )
+    halide_mutex_lock(&hexagon_cache_mutex);
+    pcache_pool temp = hexagon_cache_pool;
+    pcache_pool prev = NULL;
+
+    // Walk the free list for chunk of memory.
+    //Simplistic approach if free block size is less than the size requested assign
+    while (temp != NULL) {
+        if ((temp->used == false) &&
+            (size == temp->bytes)) {
+            temp->used = true;
+            halide_mutex_unlock(&hexagon_cache_mutex);
+            return (void *)temp->l2memory;
+        }
+        prev = temp;
+        temp = temp->next;
+    }
+
+    // If we are still here that means temp was null.
+    temp = (pcache_pool) malloc(sizeof(hexagon_cache_pool_t));
+    if (temp == NULL) {
+        halide_print(user_context, "Hexagon Cache Pool Allocation Failed.\n");
+        halide_mutex_unlock(&hexagon_cache_mutex);     
+        return NULL;
+    }
+    uint8_t *mem = (uint8_t *)HAP_cache_lock(sizeof(char) * size, NULL);
+    if ((mem == NULL) && retry) {
```

Comment:
I think this function would be a lot simpler if it were two functions (right now, allocation code is duplicated):

    find_free_buffer()
    if (retry && !free_buffer_found) {
        free_unused_buffers();
        find_free_buffer();
    }

Reply:
Added free_unused_buffers() function. Since find_free_buffer() logic is different in both scenarios, have not duplicated the allocation code. The logic is different because we need a pointer to previous node to which we can attach a new node.